src/librustc_typeck/check/wfcheck.rs

```diff
@@ -415,19 +404,141 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                 }
             }
 
-            let predicates = fcx.tcx.predicates_of(item_def_id).instantiate_identity(fcx.tcx);
-            let predicates = fcx.normalize_associated_types_in(item.span, &predicates);
-            this.check_where_clauses(fcx, item.span, &predicates);
+            this.check_where_clauses(fcx, item.span, item_def_id);
 
             fcx.impl_implied_bounds(item_def_id, item.span)
         });
     }
 
+    /// Checks where clauses and inline bounds.
     fn check_where_clauses<'fcx, 'tcx>(&mut self,
                                        fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
                                        span: Span,
-                                       predicates: &ty::InstantiatedPredicates<'tcx>)
+                                       def_id: DefId)
     {
+        use ty::subst::Subst;
+        use ty::Predicate;
+
+        let generics = self.tcx.generics_of(def_id);
+        let defaults = generics.types.iter().filter_map(|p| match p.has_default {
+                                                                true => Some(p.def_id),
+                                                                false => None,
+                                                        });
+        // Defaults must be well-formed.
+        for d in defaults {
+            fcx.register_wf_obligation(fcx.tcx.type_of(d), fcx.tcx.def_span(d), self.code.clone());
+        }
+
+        // Check that each default fulfills the bounds on it's parameter.
+        // We go over each predicate and duplicate it, substituting defaults in the self type.
```

Comment:
This part seems less obviously like something we want. It's a bit tricky, since of course there can be "interconnections" between defaults.

I guess that to start i'd like to state more clearly what property we are trying to enforce here. For example, are we checking that each default, considered individually, would not cause things to stop being well-formed?

Let's take `struct Foo<T=u32, U=i32> where C` (where `C` is some set of where-clauses) as our running example. We are already checking that `WF(C)`, which basically says that `forall T: C => WF(C)`. 

I guess you are then trying to check that `forall T: [i32/U] C => [i32/U] WF(C)`? That is, assuming that `U` is `i32`, but holding `T` abstract, we check that the predicates `C`  are well-formed? (And then we do a similar thing for `T`?)

(Does that make sense? Am I confused (or being confusing)?)